### DISCUSSION TITLE
Inconsistent pattern file load times

### DISCUSSION THREAD

**Discussion by raqibhayder** [explosion/spaCy#7774] [Help: Coding & Implementations] [answered]
I have two files, fruits and animals which contain patterns. There are about 11k and 25k patterns respectively in each file.

``` python
if __name__ == '__main__':
    nlp = spacy.load("./model")

    entity_ruler = EntityRuler(nlp, overwrite_ents=False)
    fruit_patterns = srsly.read_jsonl("./data/fruits.jsonl")
    animal_patterns = srsly.read_jsonl("./data/animals.jsonl")

    start = time.time()
    entity_ruler.add_patterns(fruit_patterns)
    lap = time.time()
    print(lap - start)
    entity_ruler.add_patterns(animal_patterns)
    end = time.time()
    print(end - lap)
```

Output: 
``` sh
19.690578937530518
49.90187907218933
```

These load times (in seconds) make sense to me.

---

When I load the patterns inside my FastAPI application, the load times are extremely slow. 

``` python
entity_ruler = EntityRuler(self.nlp, overwrite_ents=False)

# Load entity patterns
fruit_patterns = srsly.read_jsonl("./data/fruits.jsonl")
animal_patterns = srsly.read_jsonl("./data/animals.jsonl")

# Add loaded patterns to entity ruler
logger.info("Loading fruit patterns")
entity_ruler.add_patterns(fruit_patterns)
logger.info("Successfully loaded fruit patterns")

logger.info("Loading animal patterns")
entity_ruler.add_patterns(animal_patterns)
logger.info("Successfully loaded animal patterns")
```

``` sh
INFO 2021-04-13 14:21:55,210 [spacy_extractor.py:_entity_ruler:71] Loading fruit patterns
INFO 2021-04-13 14:30:32,481 [spacy_extractor.py:_entity_ruler:75] Successfully loaded fruit patterns
INFO 2021-04-13 14:30:32,482 [spacy_extractor.py:_entity_ruler:77] Loading animal patterns
INFO 2021-04-13 14:50:34,242 [spacy_extractor.py:_entity_ruler:81] Successfully loaded animal patterns
```

I tried disabling the `tagger` and `parser` and that does not make a lot of difference. 

---

I think I am missing something very silly somewhere. Any help would be really appreciated.

**Comment by raqibhayder** [comment]
The code below (extracted from [entityruler.py](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/entityruler.py)) is the bottleneck. I figured it out by running in debug mode in PyCharm and can see the difference in time it takes to execute the for loop below. 

- 18 seconds in one context vs 9 minutes in another for fruits
- 60 seconds in one context vs 20 minutes in another for animals

 I am unable to determine why is there a difference in execution time of the same piece of code being called from different contexts? 

``` python
            for label, pattern, ent_id in zip(
                phrase_pattern_labels,
                self.nlp.pipe(phrase_pattern_texts),
                phrase_pattern_ids,
            ):
                phrase_pattern = {"label": label, "pattern": pattern, "id": ent_id}
                if ent_id:
                    phrase_pattern["id"] = ent_id
                phrase_patterns.append(phrase_pattern)
```

**Reply by svlandeg** [reply]
Could it be the `nlp.pipe` part? Which pipes come before the `entity_ruler`? Are any of them doing IO as part of their `call` methods or something else that could be explaining this difference? And are you using exactly the same `nlp` model in both contexts?

(either way the time delay is crazy in your FastAPI application!)

**Reply by raqibhayder** [reply]
> Could it be the nlp.pipe part?

It is the `nlp.pipe` part. I was able to determine that while running in debug mode and stepping over statements. 

> Which pipes come before the entity_ruler?

No pipes. This is the only custom component that I use in my pipeline.

> And are you using exactly the same nlp model in both contexts?

Yes. 

---

My hunch is that it is not a spaCy issue, and has something to do with FastAPI underneath but while debugging the part that is slow is the `nlp.pipe`. 

Why do I think it is not a spaCy issue? If I change the entrypoint of the app, the loading speeds up to what is expected. 

**This directory structures makes `nlp.pipe` slow**
```
.
├── Dockerfile
├── README.md
├── app
│   ├── __init__.py
│   ├── db.py
│   ├── api.py          <<<<<<<<<<--(app is defined here)
│   ├── models.py
│   ├── ner
│   │   ├── __init__.py
│   │   └── spacy_extractor.py
│   └── schemas.py
├── config.py
├── data
│   ├── fruit.jsonl
│   └── animal.jsonl
├── main.py         <<<<<<<<<<--(entrypoint to the app is here, and app is imported here from api.py)
├── prestart.sh
├── requirements.txt
└── tests
    ├── __init__.py
    ├── conftest.py
    ├── fruit
    │   ├── __init__.py
    │   ├── test_if_fruit_is_edible.py
    ├── animal
    │   ├── __init__.py
    │   ├── test_if_animal_is_pettable.py
    └── test_api.py
```

**This directory structures makes `nlp.pipe` fast**
```
.
├── Dockerfile
├── README.md
├── app
│   ├── __init__.py
│   ├── db.py
│   ├── main.py          <<<<<<<<<<--(entrypoint to the app and app is defined here)
│   ├── models.py
│   ├── ner
│   │   ├── __init__.py
│   │   └── spacy_extractor.py
│   └── schemas.py
├── config.py
├── data
│   ├── fruit.jsonl
│   └── animal.jsonl
├── prestart.sh
├── requirements.txt
└── tests
    ├── __init__.py
    ├── conftest.py
    ├── fruit
    │   ├── __init__.py
    │   ├── test_if_fruit_is_edible.py
    ├── animal
    │   ├── __init__.py
    │   ├── test_if_animal_is_pettable.py
    └── test_main.py
```

By entrypoint I mean:
```python
if __name__ == '__main__':
    import uvicorn
    uvicorn.run(app, host='127.0.0.1', port=8080, log_level="info")
```

By app I mean:
```python
from fastapi import FastAPI

app = FastAPI()
```

So, why dont I just move to the directory structyre which speeds up the `nlp.pipe`?
Because I have tests that I need to run before deploying my project, and when running tests, the `nlp.pipe` is slow again.

I dont understand how starting the app from a different directory can affect `nlp.pipe`? 

@tiangolo: Adding Sebastian because this seems to be an issue that has something to do with FastAPI. I ran a very similar directory structure (the slow one) with Flask and did not have that issue. We are migrating from Flask to FastAPI and this is the only blocker.

@svlandeg : Do you have any other suggestions?

**Reply by tiangolo** [reply]
Hey! That's very strange, and I don't see a reason why that could happen.

It's also more difficult without seeing the code and seeing how you are calculating the time in the FastAPI app.

I can imagine some strange hypothesis, like if you were loading that multiple times on different dependencies for the same request, or doing computation in different places that could be affected by the multiple threads, or something that could be blocking the threads, or in some way re-starting the server several times and loading that on each restart... etc.

I think the best would be to create a very simple example app that replicates the problem and ask about it in a new FastAPI issue following the template there: https://github.com/tiangolo/fastapi/issues/new?assignees=&labels=question&template=question.md&title=  

By creating an issue there I will be able to make sure I check it and reply there at some point. And by following the steps on the template we'll be able to have a simple way to replicate and understand the problem. You might even find a solution while doing it. 🤓🤞

**Reply by raqibhayder** [reply]
HI @tiangolo,

This is still a blocker and am not being able to decipher the inconsistencies in pattern load time. I have created a new FastAPI issue [here](https://github.com/tiangolo/fastapi/issues/3253) and also created a simple example app that replicates the problem [here](https://github.com/raqib-hayder/fruits).

Thank you for your time.

**Comment by raqibhayder** [comment]
Another thought:

If I were to add the patterns to the model and save it to disk, will `nlp.pipe` be called on model loading for the patterns or are the patterns embedded in the model is some manner?

**Reply by raqibhayder** [reply]
Tried this, and the issue still persists.

**Comment by Kludex** [comment, marked as answer]
I'm still not sure what is the problem, but I'm 100% sure is not a FastAPI (or `uvicorn`, fwiw) issue.

Let me give you some context besides what is already here:
@raqib-hayder [created an issue](https://github.com/tiangolo/fastapi/issues/3253) on FastAPI, and you can experience the behavior described above. But the code doesn't reach `uvicorn` (you can check this by just removing the call to `uvicorn.run` :sweat_smile:).

More info:
- It works on Python 3.8 normally. This makes me think that is related to some requirements either [here](https://github.com/explosion/spaCy/blob/master/requirements.txt) or on [catalogue](https://github.com/explosion/catalogue/blob/master/requirements.txt).

My other theory is that there's something weird about the imports on that project, and the interaction with the packages :sweat_smile:

EDIT: I forgot to paste the logs at the point I've interrupted the execution:
```bash
(fruits) ➜  fruits git:(main) ✗ python main.py 
INFO 2021-05-20 02:43:23,430 [spacy_extractor.py:_entity_ruler:22] Starting to load patterns from disk
INFO 2021-05-20 02:43:23,430 [spacy_extractor.py:_entity_ruler:23] Adding fruit patterns
^CTraceback (most recent call last):
  File "main.py", line 1, in <module>
    from app import api
  File "/home/marcelo/Development/fruits/app/api.py", line 9, in <module>
    extractor = FruitAndAnimalEntityRecognizer(nlp)
  File "/home/marcelo/Development/fruits/app/ner/spacy_extractor.py", line 11, in __init__
    self._add_custom_components()
  File "/home/marcelo/Development/fruits/app/ner/spacy_extractor.py", line 14, in _add_custom_components
    self.nlp.add_pipe(self._entity_ruler(), name="base_entities", after="ner")
  File "/home/marcelo/Development/fruits/app/ner/spacy_extractor.py", line 24, in _entity_ruler
    entity_ruler.add_patterns(fruit_patterns)
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/spacy/pipeline/entityruler.py", line 225, in add_patterns
    phrase_pattern_ids,
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/spacy/language.py", line 826, in pipe
    for doc in docs:
  File "nn_parser.pyx", line 249, in pipe
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/spacy/util.py", line 513, in minibatch
    batch = list(itertools.islice(items, int(batch_size)))
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/spacy/language.py", line 811, in <genexpr>
    docs = (self.make_doc(text) for text in texts)
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/spacy/language.py", line 464, in make_doc
    return self.tokenizer(text)
  File "tokenizer.pyx", line 147, in spacy.tokenizer.Tokenizer.__call__
  File "doc.pyx", line 216, in spacy.tokens.doc.Doc.__init__
  File "doc.pyx", line 94, in spacy.tokens.doc._get_chunker
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/spacy/util.py", line 75, in get_lang_class
    return registry.languages.get(lang)
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/catalogue.py", line 92, in get
    from_entry_point = self.get_entry_point(name)
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/catalogue.py", line 138, in get_entry_point
    for entry_point in AVAILABLE_ENTRY_POINTS.get(self.entry_point_namespace, []):
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/importlib_metadata/__init__.py", line 311, in get
    flake8_bypass(self._warn)()
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/site-packages/importlib_metadata/__init__.py", line 272, in flake8_bypass
    is_flake8 = any('flake8' in str(frame.filename) for frame in inspect.stack()[:5])
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/inspect.py", line 1513, in stack
    return getouterframes(sys._getframe(1), context)
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/inspect.py", line 1490, in getouterframes
    frameinfo = (frame,) + getframeinfo(frame, context)
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/inspect.py", line 1460, in getframeinfo
    filename = getsourcefile(frame) or getfile(frame)
  File "/home/marcelo/anaconda3/envs/fruits/lib/python3.7/inspect.py", line 656, in getfile
    if isfunction(object):
KeyboardInterrupt
```
At this point (comparing the execution time), it was already finished on the `app/api.py`.

**Reply by adrianeboyd** [reply]
Some users have noticed performance regressions with particular versions of `importlib-metadata`. See #7628.

Try upgrading `catalogue`, which now includes a vendored older version of `importlib-metadata` that doesn't have this issue (we hope, anyway).

Or you can pin `importlib-metadata` to an earlier version, although this may cause conflicts with other packages. From what users reported (I was never able to replicate it locally), try `importlib-metadata<3.8.0`. For reference, the vendored version we chose is v3.2.0 due some additional issues with the `catalogue` test suite.

**Reply by raqibhayder** [reply]
> I'm still not sure what is the problem, but I'm 100% sure is not a FastAPI (or uvicorn, fwiw) issue.

@Kludex: **You are correct.** It is not a FastAPI or uvicorn issue. I was running Python 3.7, and given that it worked fine for you in Python 3.8, I switched to Python 3.8, and it not only works, but the pattern load speed is much faster than my expectations (< 20 seconds for 40k patterns). I still don't know what the underlying issue was that was slowing things down in Python 3.7 🧐

I have not needed to change anything else. 

Thank you all @Kludex @adrianeboyd @svlandeg, and @tiangolo for helping out. Much appreciated.
